File: back/src/main/java/com/back/domain/queue/listener/QueueEntryEventHandler.java
package com.back.domain.queue.listener;

import com.back.domain.queue.constant.QueueEventType;
import com.back.domain.queue.dto.QueueEventResponse;
import com.back.domain.queue.event.EntryAllowedEvent;
import com.back.domain.queue.event.QueueStatusEvent;
import lombok.RequiredArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
@Async
public class QueueEntryEventHandler {
    private final SimpMessagingTemplate messagingTemplate;

    @EventListener
    public void handleQueueStatusUpdated(QueueStatusEvent event) {
        QueueEventResponse<QueueStatusEvent> response =
                QueueEventResponse.of(QueueEventType.QUEUE_STATUS_UPDATED, event);

        messagingTemplate.convertAndSend(
                "/queue/schedules/%s/status".formatted(event.scheduleId()),
                response
        );
    }

    @EventListener
    public void handleEntryAllowed(EntryAllowedEvent event) {
        QueueEventResponse<EntryAllowedEvent> response =
                QueueEventResponse.of(QueueEventType.ENTRY_ALLOWED, event);

        messagingTemplate.convertAndSendToUser(
                event.userId().toString(),
                "/queue/schedules/%s/entry".formatted(event.scheduleId()),
                response
        );
    }
}

File: back/src/main/kotlin/com/back/domain/queue/listener/QueueEntryEventHandler.kt
package com.back.domain.queue.listener

import com.back.domain.queue.event.EntryAllowedEvent
import com.back.domain.queue.event.QueueErrorEvent
import com.back.domain.queue.event.QueueStatusEvent
import com.back.domain.waiting.sse.QueueSseEmitterRegistry
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component

@Component
@Async
class QueueEntryEventHandler(
    private val registry: QueueSseEmitterRegistry,
) {

    @EventListener
    fun handleQueueStatusUpdated(event: QueueStatusEvent) {
        registry.broadcastStatus(event.scheduleId, event)
    }

    @EventListener
    fun handleEntryAllowed(event: EntryAllowedEvent) {
        registry.sendEntryAllowed(event)
    }

    @EventListener
    fun handleQueueError(event: QueueErrorEvent) {
        registry.sendError(event)
    }
}
